src/libcore/num/cast.rs

```diff
@@ -0,0 +1,86 @@
+use fmt;
+
+/// Internal trait for APIs like `i32::cast`.
+#[unstable(feature = "num_cast_internals", issue = "0")]
+pub trait Cast<T>: Sized {
+    #[doc(hidden)]
+    fn cast(t: T) -> Result<Self, CastError>;
+}
+
+/// Error type returned from APIs like `i32::cast`, indicates that a cast could
+/// not be performed losslessly.
+#[unstable(feature = "num_cast", issue = "0")]
+#[derive(Debug)]
+pub struct CastError(());
```

Comment:
IDK, for me the operation feels similar to `checked_add` and that returns an option as well. Or is the policy to return `Result` everywhere for new APIs?